pyminio/main.py

```diff
@@ -354,27 +359,66 @@ def rm(self, path: str, recursive: bool = False):
         match = Match(path)
         self.minio_obj.remove_object(match.bucket, match.relative_path)
 
-    # TODO: implement copy recursive (Task #12)
-    # def _copy_directory(self, from_path: str, to_path: str):
-    #     objects = self.get_objects_at(from_path)
-    #     files = []
-    #     directories = []
-    #     for obj in objects:
-    #         if obj.is_dir:
-    #             directories.append(obj)
-    #         else:
-    #             files.append(obj)
-    #     if directories == []:
-    #         pass  # make path
-    #         # self.mkdirs(join(to_path, dirname(from_path), ''))
-    #     else:
-    #         #  recursive run with next directories
-    #         for directory in directories:
-    #             self._copy_directory(
-    #                 from_path=,
-    #                 to_path=join(to_path, basename(directory))
-    #     for file_to_copy in files:
-    #         #  create file
+    def _get_destination(self, from_path: str, to_path: str):
+        from_match = Match(from_path)
+        to_match = Match(to_path)
+
+        if from_match.is_file():
+            return Match.infer_operation_destination(from_match, to_match)
+
+        if to_match.is_dir():
+            if self.exists(to_match.path):
+                return Match(join(to_match.path, basename(
+                    join(from_match.bucket, from_match.prefix)[:-1]), ''))
+
+        else:
+            raise ValueError(
+                "can not activate this method from directory to a file.")
+
+        return to_match
+
+    @_validate_directory
+    def copy_recursively(self, from_path: str, to_path: str):
+        """Copy recursively the content of from_path, to to_path.
+
+        If you acctually wanted to copy from_path as a folder,
+        add that folder's name to to_path and that new path
+        will be created for you.
+
+        Args:
+            from_path: source path to a file.
+            to_path: destination path.
+            recursive: copy content recursively.
+        """
+        files_to_copy = []
+        dirs_to_copy = [from_path]
+
+        while len(dirs_to_copy) > 0:
+            current_dir_match = Match(dirs_to_copy.pop(0))
+            objects_in_directory = self._get_objects_at(current_dir_match)
+            dirs = []
+
+            for obj in objects_in_directory:
+                obj_path = f"/{obj.bucket_name}/{obj.object_name}"
+                if obj.is_dir:
+                    dirs.append(obj_path)
+                else:
+                    files_to_copy.append(AttrDict(
+                        from_path=obj_path,
+                        to_path=join(to_path, obj_path.replace(from_path, '')),
+                    ))
```

Comment:
y not just doing the cp here?
and also creating the needed directory before?
maybe add a feature in cp to create missing directories

Reply:
i didnt cp there because the directory is not yet existing, and i done this to decrees requests.
so im making the directories and then adding the files.
although i might not need to make the directories before copying the files. need further reaserch